nodes/recipe_handler.py

```diff
@@ -199,47 +226,73 @@ def stop_recipe_service(self, data):
 
     def register_services(self):
         """Register services for instance"""
-        rospy.Service(services.START_RECIPE, StartRecipe, self.start_recipe_service)
+        rospy.Service(services.START_RECIPE, StartRecipe, 
+            self.start_recipe_service)
         rospy.Service(services.STOP_RECIPE, Empty, self.stop_recipe_service)
         rospy.set_param(
             params.SUPPORTED_RECIPE_FORMATS,
             ','.join(RECIPE_INTERPRETERS.keys())
         )
         return self
 
-    def recover_any_previous_recipe(self, environment):
+    def recover_any_previous_recipe(self):
         """
         Attempt to resume any previous recipe that was started but
         not completed.
         """
         # Get the recipe that has been started most recently
         start_view = self.env_data_db.view(
             "openag/by_variable",
-            startkey=[environment, "desired", RECIPE_START.name],
-            endkey=[environment, "desired", RECIPE_START.name, {}],
+            startkey=[self.environment, "desired", RECIPE_START.name],
+            endkey=[self.environment, "desired", RECIPE_START.name, {}],
             group_level=3
         )
         if len(start_view) == 0:
+            trace("recipe_handler: No previous recipe to recover.")
             return
         start_doc = start_view.rows[0].value
+        trace("recipe_handler: start_doc=%s", start_doc)
         # If a recipe has been ended more recently than the most recent time a
         # recipe was started, don't run the recipe
         end_view = self.env_data_db.view(
             "openag/by_variable",
-            startkey=[environment, "desired", RECIPE_END.name],
-            endkey=[environment, "desired", RECIPE_END.name, {}],
+            startkey=[self.environment, "desired", RECIPE_END.name],
+            endkey=[self.environment, "desired", RECIPE_END.name, {}],
             group_level=3
         )
         if len(end_view):
             end_doc = end_view.rows[0].value
+            trace("recipe_handler: end_doc=%s", end_doc)
             if (end_doc["timestamp"] > start_doc["timestamp"]):
+                trace("recipe_handler: RETURNING: end_time=%s > start_time=%s", 
+                    end_doc["timestamp"], start_doc["timestamp"])
                 return
         # Run the recipe
+        trace("recipe_handler: restarting recipe=%s at time=%s", 
+            start_doc["value"], start_doc["timestamp"])
         self.start_recipe_service(
             StartRecipe._request_class(start_doc["value"]),
             start_doc["timestamp"]
         )
-
+        
+    def save_recipe_dp(self, variable):
```

Comment:
Sure, if that is the best way to go.  The way I have it implemented now, it won't put a _second_ recipe_start into the DB if it is a restart situation.  (but the message is still published).  Not sure if we want logic like that in a separate node?